packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js

```diff
@@ -22,58 +24,74 @@ let useSyncExternalStoreWithSelector;
 let use;
 let PropTypes;
 let textCache;
-let window;
-let document;
 let writable;
 let CSPnonce = null;
-let container;
+let container = null;
 let buffer = '';
 let hasErrored = false;
 let fatalError = undefined;
+let originalDocument;
+let originalWindow;
+
+function resetModules() {
+  jest.resetModules();
+  ({JSDOM, VirtualConsole: JSDOMVirtualConsole} = require('jsdom'));
+  Scheduler = require('scheduler');
+  React = require('react');
+  ReactDOMClient = require('react-dom/client');
+  ReactDOMFizzServer = require('react-dom/server');
+  Stream = require('stream');
+  Suspense = React.Suspense;
+  if (gate(flags => flags.enableSuspenseList)) {
+    SuspenseList = React.SuspenseList;
+    use = React.experimental_use;
+  }
 
-describe('ReactDOMFizzServer', () => {
-  beforeEach(() => {
-    jest.resetModules();
-    JSDOM = require('jsdom').JSDOM;
-    Scheduler = require('scheduler');
-    React = require('react');
-    ReactDOMClient = require('react-dom/client');
-    ReactDOMFizzServer = require('react-dom/server');
-    Stream = require('stream');
-    Suspense = React.Suspense;
-    if (gate(flags => flags.enableSuspenseList)) {
-      SuspenseList = React.SuspenseList;
-      use = React.experimental_use;
-    }
+  PropTypes = require('prop-types');
 
-    PropTypes = require('prop-types');
+  if (gate(flags => flags.source)) {
+    // The `with-selector` module composes the main `use-sync-external-store`
+    // entrypoint. In the compiled artifacts, this is resolved to the `shim`
+    // implementation by our build config, but when running the tests against
+    // the source files, we need to tell Jest how to resolve it. Because this
+    // is a source module, this mock has no affect on the build tests.
+    jest.mock('use-sync-external-store/src/useSyncExternalStore', () =>
+      jest.requireActual('react'),
+    );
+  }
+  useSyncExternalStore = React.useSyncExternalStore;
+  useSyncExternalStoreWithSelector = require('use-sync-external-store/with-selector')
+    .useSyncExternalStoreWithSelector;
+}
+
+function resetJSDOM(markup) {
+  // Test Environment
+  const virtualConsole = new JSDOMVirtualConsole();
+  virtualConsole.sendTo(console, {
+    omitJSDOMErrors: true,
+  });
+  virtualConsole.on('jsdomError', error => {
+    console.error(error);
+  });
```

Comment:
@eps1lon what are these for? I'm not well versed in jsdom-isms but it's confusing you need to omit jsdom errors in the sendTo and then you rewrite them manually to the console. Does it always route to .log and you want .error instead?

Reply:
So if I understood these correctly, they're not actually omitting the errors. Just ensuring that they're forwarded in a way that is more "idiomatic". Without that, `jsdom` will issue `console.errors` with multiple arguments, but no placeholders. Thus triggering https://github.com/facebook/react/blob/3de926449681015c4f1eec1147baa03a3cb823f5/scripts/jest/matchers/toWarnDev.js#L236-L248

What we're doing here, is just recreating the JSDOM environment with the given HTML i.e. this is how `jest-environment-jsdom` already works: https://github.com/facebook/jest/blob/e21c5aba950f6019bbfde2f8233ac96d1fcaef42/packages/jest-environment-jsdom/src/index.ts#L41-L45